File: android/src/main/java/com/margelo/nitro/nitroandroidpip/NitroAndroidPip.kt
package com.margelo.nitro.nitroandroidpip

import android.os.Build
import android.util.Log
import com.facebook.proguard.annotations.DoNotStrip

@DoNotStrip
class NitroAndroidPip : HybridNitroAndroidPipSpec() {

    companion object {
        private const val TAG = "NitroAndroidPip"
    }

    override fun setPipOptions(options: IPipOptions?, actions: Array<IPipAction>?) {
        val list = actions?.toList()?.take(3) ?: emptyList()
        NitroAndroidPipManager.setPipOptions(options, list)
    }

    override fun startPip() {
        Log.d(TAG, "startPip() called")
        NitroAndroidPipManager.startPip()
    }

    override fun stopPip() {
        Log.d(TAG, "stopPip() called")
        NitroAndroidPipManager.stopPip()
    }

    override fun isPipSupported(): Boolean {
        return NitroAndroidPipManager.isPipSupported()
    }

    override fun isPipActive(): Boolean {
        return NitroAndroidPipManager.isPipActive()
    }

    override fun addPipListener(callback: (isPipActive: Boolean) -> Unit): Unit {
        // Wrap the JS callback to catch exceptions; the manager will invoke on main thread.
        val safeCallback: (Boolean) -> Unit = { active ->
            try {
                callback(active)
            } catch (e: Exception) {
                Log.e(TAG, "Error executing JS PiP callback", e)
            }
        }
        NitroAndroidPipManager.setListener(safeCallback)
    }

    override fun removePipListener(): Unit {
        NitroAndroidPipManager.clearListener()
    }
}

File: ios/NitroAndroidPip.swift
import Foundation
import NitroModules
import OSLog

class NitroAndroidPip: HybridNitroAndroidPipSpec {

    private let logger = Logger(subsystem: "com.margelo.nitro.nitroandroidpip", category: "NitroAndroidPip")

    public func setPipOptions(options: IPipOptions?, actions: [IPipAction]?) throws -> Void {
        // iOS: Android-specific functionality; no-op
        logger.info("[NitroAndroidPip] setPipOptions called on iOS - no-op")
    }

    public func startPip() throws -> Void {
        logger.info("[NitroAndroidPip] startPip called on iOS - no-op")
    }

    public func stopPip() throws -> Void {
        logger.info("[NitroAndroidPip] stopPip called on iOS - no-op")
    }

    public func isPipSupported() throws -> Bool {
        // iOS-side of this Android-specific module: report false
        return false
    }

    public func isPipActive() throws -> Bool {
        return false
    }

    public func addPipListener(callback: @escaping (Bool) -> Void) throws -> Void {
        logger.info("[NitroAndroidPip] addPipListener called on iOS - invoking callback(false) and no-op")
        DispatchQueue.main.async {
            callback(false)
        }
    }

    public func removePipListener() throws -> Void {
        logger.info("[NitroAndroidPip] removePipListener called on iOS - no-op")
    }
}
